examples/ucp_hello_world_multi_ctx.c

```diff
@@ -0,0 +1,712 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2001-2016.  ALL RIGHTS RESERVED.
+* Copyright (C) Advanced Micro Devices, Inc. 2018. ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#ifndef HAVE_CONFIG_H
+#  define HAVE_CONFIG_H /* Force using config.h, so test would fail if header
+                           actually tries to use it */
+#endif
+
+/*
+ * UCP hello world client / server example utility
+ * -----------------------------------------------
+ *
+ * Server side:
+ *
+ *    ./ucp_hello_world
+ *
+ * Client side:
+ *
+ *    ./ucp_hello_world -n <server host name>
+ *
+ * Notes:
+ *
+ *    - Client acquires Server UCX address via TCP socket
+ *
+ *
+ * Author:
+ *
+ *    Ilya Nelkenbaum <ilya@nelkenbaum.com>
+ *    Sergey Shalnov <sergeysh@mellanox.com> 7-June-2016
+ */
+
+#include "hello_world_util.h"
+
+#include <ucp/api/ucp.h>
+
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <sys/epoll.h>
+#include <netinet/in.h>
+#include <assert.h>
+#include <netdb.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>  /* getopt */
+#include <ctype.h>   /* isprint */
+#include <pthread.h> /* pthread_self */
+#include <errno.h>   /* errno */
+#include <time.h>
+#include <signal.h>  /* raise */
+
+#define MAX_UCP_CTXS    16
+
+struct msg {
+    uint64_t        data_len;
+};
+
+struct ucx_context {
+    int             completed;
+};
+
+typedef struct ucx_mem_type_ctx {
+    ucp_context_h ucp_context;
+    ucp_worker_h ucp_worker;
+    ucp_address_t *local_addr;
+    ucp_address_t *peer_addr;
+    size_t local_addr_len;
+    size_t peer_addr_len;
+    int ordinal;
+} ucx_mem_type_ctx_t;
+
+enum ucp_test_mode_t {
```

Comment:
seems it can be unitied with existing `ucp_hello_world` by adding new parameter.

Reply:
@brminich considered this but leaving this as a separate test as it still needs to be updated with the following variations from the TODO list:
-   Need to add any-to-any
-   Need to add concurrent use of contexts

After these additions, if it can be generalized to have the basic case as well, I'll try to unify with it.